### PR TITLE
*: re-implement partition pruning for better performance (#14679)

### PR DESCRIPTION
**Pull request by tiancaiamao**
Cherry-pick https://github.com/pingcap/tidb/pull/14679 to 3.0.11, make range pruning as fast as possible

Before:
79076815 ns/op
22.99% of CPU cost
image

After 
251407 ns/op
1.09% of CPU cost
image

### CODE REVIEW

**Comment by tiancaiamao** [comment]
PTAL @jackysp

**Comment by shenli** [comment]
can you compare it with 3.0.5?

**Comment by tiancaiamao** [comment]
v3.0.5 The bench result of same case is 19620187 ns/op

There is no cost in `ParseSimpleExprsWithSchema` before https://github.com/pingcap/tidb/pull/13490, the flame graph is totally different:


image

**Comment by tiancaiamao** [comment]
I have to take a rest, when I wake up I'll continue with it and fix the CI.

**Comment by tiancaiamao** [comment]
The CI hang...

**Comment by jackysp** [comment]
PTAL @lysu @imtbkcat

**Comment by qrr1995** [comment]
/run-all-tests pd=release-3.0 tikv=release-3.0 tidb-test=release-3.0

File: table/tables/partition.go [hunk 1]
```diff
@@ -125,6 +129,43 @@ type PartitionExpr struct {
 	UpperBounds []expression.Expression
 	// Expr is the hash partition expression.
 	Expr expression.Expression
+
+	// The new range partition pruning
+	*ForRangePruning
```
Comment (lysu) [inline table/tables/partition.go]: this is an optimization that avoids calculate `lessThenData` every time but doesn't exists in master branch?

**Comment by lysu** [comment]
new

what's workload running when capturing this graph? why so many `makeUnionAllChildren` but maybe we should see `partitionRangeForCNFExpr`?

**Comment by tiancaiamao** [comment]
> new
> 
> what's workload running when capturing this graph? why so many `makeUnionAllChildren` but maybe we should see `partitionRangeForCNFExpr`?

https://github.com/pingcap/tidb/pull/15368/files#diff-1b0a7f60f2fea2fa689b1dfda250bc3dR1475

**Comment by lysu** [comment]
> > new
> > what's workload running when capturing this graph? why so many `makeUnionAllChildren` but maybe we should see `partitionRangeForCNFExpr`?
> 
> https://github.com/pingcap/tidb/pull/15368/files#diff-1b0a7f60f2fea2fa689b1dfda250bc3dR1475

my mistake..- -

File: table/tables/partition.go (hunk 1 above)
Comment (tiancaiamao) [reply table/tables/partition.go]: Yes, it's a TODO here https://github.com/pingcap/tidb/pull/15169/files#diff-0411f48b4045359c9f116ef6a779c3efR331

And another tiny change is lessThanData could be int/uint when I find a CI case fail.

**Comment by tiancaiamao** [comment]
/run-all-tests pd=release-3.0 tikv=release-3.0 tidb-test=release-3.0

**Comment by lysu** [comment]
LGTM

**Comment by qrr1995** [comment]
/run-all-tests pd=release-3.0 tikv=release-3.0 tidb-test=release-3.0

**Comment by tiancaiamao** [comment]
Close as I've got the binary, this commit finish its mission.

**Pull request closed without merging**